test/s3_test.go

```diff
@@ -0,0 +1,255 @@
+					expectPassRead:  false,
+					expectPassWrite: false,
+				},
+			},
+		},
+		{
+			testName: "TestBucket2",
+			vars: map[string]interface{}{
+				"test_bucket_name": "",
+				"read_only_paths":  []string{"other/path/to/ro-folder"},
+				"read_write_paths": []string{"other/path/to/rw-folder"},
+			},
+			objTestCases: []ObjectTestCase{
+				{
+					key:             "other/path/to/ro-folder/obj1",
+					encryption:      "AES256",
+					expectPassRead:  true,
+					expectPassWrite: false,
+				},
+				{
+					key: "path/to/rw-folder/obj2",
+					// not setting encryption here to make sure we cannot upload unencrypted objects
+					encryption:      "",
+					expectPassRead:  false,
+					expectPassWrite: false,
+				},
+			},
+		},
+	}
+}
+
+func TestTerraformS3Module(t *testing.T) {
+
+	// os.Setenv("SKIP_", "true")
+	// os.Setenv("TERRATEST_REGION", "us-east-1")
+
+	// For convenience - uncomment these as well as the "os" import
+	// when doing local testing if you need to skip any sections.
+	//
+	// The good usage for skipping stages is local testing, when you want to make sure your test code is fine,
+	// and not waste time building new infrastructure every time.
+	// A nice approach for doing that would be:
+	// 1- Make sure there are no local files in the dirs (`rm -rf .test-dir terraform.tfstate*`)
+	// 2- Skip teardown* and run tests. (create and keep infrastructure, tfstate will be into local folders)
+	// 3- In your next run skip `pick*` steps (which makes sure random values in Terratest state won't be updated)
+	//    obs: if you forget and run tests once, you can make use of `terraform.tfstate.backup` file to restore it
+	// 4- Run local Tests; Do your thing;
+	// 5- When you are done and want to destroy the infrastructure -> comment back teardown* steps
+	// 6- Comment back pick* steps.
+
+	// os.Setenv("SKIP_pick_new_randoms", "true")
+	// os.Setenv("SKIP_setup_options", "true")
+	// os.Setenv("SKIP_create_bucket", "true")
+	// os.Setenv("SKIP_validate_bucket", "true")
+	// os.Setenv("SKIP_create_ro_objects", "true")
+	// os.Setenv("SKIP_setup_role_options", "true")
+	// os.Setenv("SKIP_create_role", "true")
+	// os.Setenv("SKIP_validate_bucket_and_policies", "true")
+
+	// Warning: if you skip these steps, Terratest state will be stored in local folder under .test-data inside each TF dir you skip.
+	// These are purposedly stored in the known TF dir (not tmp folder) to make sure you can re-run tests on them at any time.
+	// Remember removing those folders after you finish your tests so that it won't affect the next time you run local tests.
+	// os.Setenv("SKIP_teardown", "true")
+	// os.Setenv("SKIP_teardown_role", "true")
+	//
+	// string to be used as body for files to be created
+	const test_body = "test"
+
+	// list of different buckets that will be created to be tested
+	bucketTestCases := initTestCases()
+
+	// When using test_structure functions + parallel tests with random IDs we will run into consistency problems.
```

Comment:
I agree that when trying to do local testing, it can be confusing and prone to human error. But it also does facilitate and speeds up the work when you're dealing with infrastructure that takes 20+ minutes to be created and destroyed (EMR, RDS, ES).
 
Another use case is when trying out the test code itself. It probably won't be updated often so I agree we may not need it for that purpose.

Though I think having a separated branch (e.g. `local-tests`) containing those more complex skips and verifications could be useful for when it's needed. The `README.md` inside `test` could add the guidelines... something like:

> If you wish to run local testing without re-deploying infrastructure: 
> 1 - check into branch `local-tests`; 
> 2- merge your work; 
> 3 - run your local tests using skip vars;

So I could push the current code there and simplify the code in this main branch. What do you think?

Reply:
Simplified code and comments in here https://github.com/hbarros-caylent/terraform-aws-s3/commit/fb3a0a3b4983ff6b7f5d6ebff51242a1ca3b6da9 . Let me know if the `local-tests` branch is a good idea so I can push that one in here too. (I pushed to my fork here https://github.com/hbarros-caylent/terraform-aws-s3/tree/local-tests)